File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.java
class BSTIterator {
    List<Integer> vals;
    int idx;

    public BSTIterator(TreeNode root) {
        vals = new ArrayList<>();
        idx = -1;
        iterate(root);
    }

    private void iterate(TreeNode node) {
        if (node.left != null) {
            iterate(node.left);
        }
        vals.add(node.val);
        if (node.right != null) {
            iterate(node.right);
        }
    }
    
    public int next() {
        idx++;
        return vals.get(idx);
    }
    
    public boolean hasNext() {
        if (idx < vals.size() - 1) {
            return true;
        }
        return false;
    }
}

File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.py
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.vals = list()
        self.iterate(root)
        self.idx = -1


    def iterate(self, node):
        if node.left:
            self.iterate(node.left)
        self.vals.append(node.val)
        if node.right:
            self.iterate(node.right)


    def next(self) -> int:
        self.idx += 1
        return self.vals[self.idx]


    def hasNext(self) -> bool:
        return True if self.idx < len(self.vals) - 1 else False
